### PR TITLE
fixes #4204 - add config groups feature - assign multiple config groups to a host or hostgroup

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by ares** [comment]
Could you please rename PuppetGroup to something more generic? Maybe ConfigGroup, ConfigRole or something that would be reusable for Chef (and others)

**Comment by isratrade** [comment]
@ares, good idea. will rename PuppetGroup.  I like ConfigGroup.  Any other ideas for names?

**Comment by GregSutcliffe** [comment]
+1 For ConfigGroup

**Comment by ohadlevy** [comment]
+1

**Comment by isratrade** [comment]
@ares, changed to ConfigGroup.  What about the name 'classes'?  Is this generic enough?  or should be recipes? modules? items?

**Comment by isratrade** [comment]
currently one `config_group` has_many `config_group_classes`

**Comment by GregSutcliffe** [comment]
+1 for Classes

**Comment by ares** [comment]
I think we should leave classes because for chef, we may need different object representing cookbook (module) or recipe (class), since there are differences (order in runlist etc). I believe that ConfigGroup is more like role in chef terminology. So +1 from me.

Btw we don't plan to support nesting of config groups atm, do we? That is meant to be done by hostgroup nesting?

**Comment by ohadlevy** [comment]
yeah, I dont expect nesting, its just there so you dont have to define the
same set of classes again and again.

**Comment by isratrade** [comment]
@ares, no nesting for config_groups.

**Comment by ares** [comment]
sure, I was just curious ;-) I'm looking forward to the merge of this :+1:

File: app/views/config_groups/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<% title _("Edit Config group") %>
+
+<%= render 'form' %>
```
Comment (GregSutcliffe) [inline app/views/config_groups/edit.html.erb:3]: I don't see a _form partial in this patch.

File: app/views/config_groups/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<% title _("New Config group") %>
+
+<%= render 'form' %>
```
Comment (GregSutcliffe) [inline app/views/config_groups/new.html.erb:3]: Here too, no _form in views/config_groups

**Comment by ohadlevy** [comment]
@isratrade mind having a look at the test failure? thanks!

File: app/assets/javascripts/class_edit.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
+//= require underscore
```
Comment (ohadlevy) [inline app/assets/javascripts/class_edit.js]: are we using underscore here? I wasn't sure...
if not, i guess its makes more sense to do it in a separate commit

File: app/assets/javascripts/class_edit.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  links.attr('data-original-title', __('Click to undo adding this class'));
+  links.tooltip();
+  link.removeClass('glyphicon-plus-sign').addClass('glyphicon-minus-sign');
+
+  $('#selected_classes').append(content);
+
+  $("#selected_puppetclass_"+ id).show('highlight', 5000);
+  $("#puppetclass_"+ id).addClass('selected-marker').hide();
+
+  $(document.body).trigger('AddedClass', link);
```
Comment (ohadlevy) [inline app/assets/javascripts/class_edit.js:37]: nitpick: some docs around what is this trigger would be helpful

File: app/assets/stylesheets/application.scss [hunk 5]
```diff
@@ -340,3 +345,12 @@ table.table thead .sorting_desc_disabled { background: #f9f9f9 url('sort_desc_di
   top: 30px;
   right: 30px;
 }
+
+.config_group, .selected_config_group   {
+  min-height: 30px;
+}
+
+.strikethrough {
+  //text-decoration: line-through;
```
Comment (ohadlevy) [inline app/assets/stylesheets/application.scss]: consider rename? and remove commented out code

File: app/controllers/api/v2/config_groups_controller.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+module Api
+  module V2
+    class ConfigGroupsController < V2::BaseController
+      #before_filter :find_resource
```
Comment (ohadlevy) [inline app/controllers/api/v2/config_groups_controller.rb]: commented? does it work?

File: app/models/host/managed.rb [hunk 7]
```diff
@@ -6,6 +6,12 @@ class Host::Managed < Host::Base
 
   has_many :host_classes, :dependent => :destroy, :foreign_key => :host_id
   has_many :puppetclasses, :through => :host_classes
+
+  def puppetclasses_with_groups
+    ids = (host_classes.pluck(:puppetclass_id) + config_group_classes.pluck(:puppetclass_id)).uniq
+    Puppetclass.where(:id => ids)
```
Comment (ohadlevy) [inline app/models/host/managed.rb]: I wonder if inner selects would be better here, e.g.

```
puppetclass.where(id host_classes.puppetclasses OR config_group_classes.puppetclass) 
```

File: app/assets/javascripts/class_edit.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  $('#selected_config_group_' + id).children('a').tooltip('hide');
+  $('#selected_config_group_' + id).remove();
+
+  var puppetclasses =  $(item).attr('data-puppetclass_ids')
+  var puppetclass_ids = $.parseJSON(puppetclasses);
+  console.log('trying to remove ' + puppetclass_ids);
+  $.each(puppetclass_ids, function(index,puppetclass_id){
+    var pc = $('#selected_puppetclass_' + puppetclass_id)
+    var pc_link = $("a[data-class-id='" + puppetclass_id + "']");
+    if ( (pc_link.size() > 0) && (pc.size() > 0) && (!(_.contains(inherited_ids, puppetclass_id)))  ) {
```
Comment (ohadlevy) [inline app/assets/javascripts/class_edit.js]: inherited_ids is unresolved?

File: app/assets/javascripts/class_edit.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+      }
+    }
+  })
+}
+
+function removeConfigGroup(item){
+  var id = $(item).attr('data-group-id');
+  $('#config_group_' + id).removeClass('selected-marker').show();
+  $('#selected_config_group_' + id).children('a').tooltip('hide');
+  $('#selected_config_group_' + id).remove();
```
Comment (ohadlevy) [inline app/assets/javascripts/class_edit.js:117]: nitpick:  Duplicated jQuery selector, which is less efficient.

Comment (ohadlevy) [inline app/assets/javascripts/class_edit.js:117]: and in a few other places

File: app/controllers/hostgroups_controller.rb [hunk 10]
```diff
@@ -88,7 +88,7 @@ def environment_selected
 
     @hostgroup ||= Hostgroup.new
     @hostgroup.environment = @environment if @environment
-    render :partial => 'puppetclasses/class_selection', :locals => {:obj => (@hostgroup)}
+    render :partial => 'puppetclasses/class_selection', :locals => {:obj => (@hostgroup), :type => 'hostgroup'}
```
Comment (ohadlevy) [inline app/controllers/hostgroups_controller.rb:91]: why do you need to pass the type too? you can ask the obj for its type, maybe can move into a helper

File: app/controllers/puppetclasses_controller.rb [hunk 11]
```diff
@@ -69,12 +69,12 @@ def get_host_or_hostgroup
           @obj.type = "Host::Managed"
         end
         # puppetclass_ids is removed since it causes an insert on host_classes before form is submitted
```
Comment (ohadlevy) [inline app/controllers/puppetclasses_controller.rb]: maybe update comment too

File: app/views/config_groups/_selected_config_group.html.erb [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+  </span>
+  <%= link_to(' Remove', '#', :'data-puppetclass_ids' => config_group.puppetclass_ids.to_json, :'data-group-id' => config_group.id, :'data-id' => config_group.id, :'data-type' => type, :onclick => 'removeConfigGroup(this)', :rel=>'twipsy', :class => 'btn btn-default btn-sm glyphicon glyphicon-minus-sign') unless obj.parent_config_groups.include?(config_group) %>
+  <ul class='hide' id='<%= "cgroup_" + config_group.id.to_s %>'>
+    <% config_group.puppetclasses.each do |cg_puppetclass|%>
+          <% strike = (obj.environment && obj.environment.puppetclasses.map(&:id).include?(cg_puppetclass.id) ? '' : 'strikethrough') %>
+          <% tip = strike == 'strikethrough' ? _('not in environment') : '' %>
+          <li>
+            <%= content_tag(:span, :class => strike, :'data-original-title' => tip, :rel => "twipsy") do %>
+              <%= cg_puppetclass.name + ' (' + cg_puppetclass.id.to_s + ')'%>
```
Comment (ohadlevy) [inline app/views/config_groups/_selected_config_group.html.erb]: remove the id?

File: app/views/config_groups/_selected_config_group.html.erb [hunk 13]
```diff
@@ -0,0 +1,17 @@
+<%= content_tag_for :li, config_group, :selected, :class => cycle('even', 'odd') do %>
+  <span>
+   <%= link_to(image_tag("bullet_toggle_plus.png") + " " + config_group.name, '#', :'data-puppetclass_ids' => config_group.puppetclass_ids.to_json, :'data-group-id' => config_group.id, :'data-id' => config_group.id, :'data-type' => type, :onclick => "$('#cgroup_#{config_group.id}').fadeToggle('slow')", :rel=>'twipsy', :class => 'config_group_name', :'data-original-title' => (_('included already from parent') if obj.parent_config_groups.include?(config_group)) )  %>
```
Comment (ohadlevy) [inline app/views/config_groups/_selected_config_group.html.erb]: hard to read, maybe convert to a data hash?

``` ruby
<%= link_to(image_tag("bullet_toggle_plus.png") + " " + config_group.name, '#',
               :data    => {
                   :puppetclass_ids => config_group.puppetclass_ids,
                   :'group-id'      => config_group.id,
```

File: app/views/config_groups/_selected_config_group.html.erb [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+  </span>
+  <%= link_to(' Remove', '#', :'data-puppetclass_ids' => config_group.puppetclass_ids.to_json, :'data-group-id' => config_group.id, :'data-id' => config_group.id, :'data-type' => type, :onclick => 'removeConfigGroup(this)', :rel=>'twipsy', :class => 'btn btn-default btn-sm glyphicon glyphicon-minus-sign') unless obj.parent_config_groups.include?(config_group) %>
+  <ul class='hide' id='<%= "cgroup_" + config_group.id.to_s %>'>
+    <% config_group.puppetclasses.each do |cg_puppetclass|%>
+          <% strike = (obj.environment && obj.environment.puppetclasses.map(&:id).include?(cg_puppetclass.id) ? '' : 'strikethrough') %>
```
Comment (ohadlevy) [inline app/views/config_groups/_selected_config_group.html.erb]: 1. can move to a helper?
2. going to be lots of queries for large data set

File: app/views/config_groups/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+<table class="table table-bordered table-striped table-two-pane">
+  <tr>
+    <th><%= sort :name %></th>
+    <th><%=_('Classes') %></th>
+    <th></th>
+  </tr>
+  <% @config_groups.each do |config_group| %>
+    <tr>
+      <td><%= link_to_if_authorized h(config_group.name), hash_for_edit_config_group_path(config_group)%></td>
+      <td style="max-width:600px;"><%= config_group.puppetclasses.pluck(:name).to_sentence %> </td>
```
Comment (ohadlevy) [inline app/views/config_groups/index.html.erb]: not sure if it make sense to list the classes, style should not be here regardless.

File: app/views/config_groups/index.html.erb [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+<table class="table table-bordered table-striped table-two-pane">
+  <tr>
+    <th><%= sort :name %></th>
+    <th><%=_('Classes') %></th>
+    <th></th>
+  </tr>
+  <% @config_groups.each do |config_group| %>
+    <tr>
+      <td><%= link_to_if_authorized h(config_group.name), hash_for_edit_config_group_path(config_group)%></td>
```
Comment (ohadlevy) [inline app/views/config_groups/index.html.erb]: h is not required

File: app/assets/javascripts/class_edit.js (hunk 3 above)
Comment (isratrade) [inline app/assets/javascripts/class_edit.js]: @ohadlevy, I was going to put it in a separate commit in application.js, but then I decided to just include it in class_edit.js for now, since this is currently the only file that uses it.

File: app/views/puppetclasses/_classes_in_groups.html.erb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+<%= content_tag_for :li, klass, :selected, :class => cycle('even', 'odd') do %>
+  <%= link_to klass.name, "#", :'data-original-title' => _('belongs to config group'), :onclick => "", :rel => "twipsy",
```
Comment (ohadlevy) [inline app/views/puppetclasses/_classes_in_groups.html.erb]: why define onclick?

File: db/migrate/20140226095754_rename_to_config_group.rb [hunk 18]
```diff
@@ -0,0 +1,7 @@
+class RenameToConfigGroup < ActiveRecord::Migration
+  def change
+    rename_table :puppet_groups, :config_groups
```
Comment (ohadlevy) [inline db/migrate/20140226095754_rename_to_config_group.rb]: can we squash the migrations?

File: app/controllers/api/v2/config_groups_controller.rb (hunk 6 above)
Comment (isratrade) [inline app/controllers/api/v2/config_groups_controller.rb]: @ohadlevy, will add the remaining CRUD in API v2.  I just added index temporarily since I used it for Angular.

File: config/routes.rb [hunk 19]
```diff
@@ -147,6 +147,10 @@
     end
   end
 
+  resources :config_groups, :except => [:show] do
+    get 'auto_complete_search', :on => :collection
```
Comment (ohadlevy) [inline config/routes.rb:151]: missing permissions?

File: app/models/host/managed.rb (hunk 7 above)
Comment (isratrade) [inline app/models/host/managed.rb]: @ohadlevy, agreed.  Will refactor.

File: app/controllers/hostgroups_controller.rb (hunk 10 above)
Comment (isratrade) [inline app/controllers/hostgroups_controller.rb:91]: I felt it was cleaner to pass `type` explictly, rather than having `obj.class.name.tableize.singularize` in a bunch of places.

File: app/views/config_groups/_selected_config_group.html.erb (hunk 12 above)
Comment (isratrade) [inline app/views/config_groups/_selected_config_group.html.erb]: @ohadlevy, yes, will remove, it was just there for debugging.

File: app/views/config_groups/index.html.erb (hunk 15 above)
Comment (isratrade) [inline app/views/config_groups/index.html.erb]: @ohadlevy, if I didn't add max-width, if there are a lot of puppletclasses, then the right column with the edit button got squashed and the button drop-down arrow moved to the next line. @abenari ?

File: app/assets/javascripts/class_edit.js [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+  var link = content.children('a');
+  var links = content.find('a');
+  links.attr('onclick', 'remove_puppet_class(this)');
+  links.attr('data-original-title', __('Click to undo adding this class'));
+  links.tooltip();
+  link.removeClass('glyphicon-plus-sign').addClass('glyphicon-minus-sign');
+
+  $('#selected_classes').append(content);
+
+  $("#selected_puppetclass_"+ id).show('highlight', 5000);
```
Comment (ohadlevy) [inline app/assets/javascripts/class_edit.js:33]: if we do this often, i suggest to move to a helper function, DRY

Comment (ohadlevy) [inline app/assets/javascripts/class_edit.js:33]: in general, add group and add class seems very similar :)

File: db/migrate/20140226095754_rename_to_config_group.rb (hunk 18 above)
Comment (isratrade) [inline db/migrate/20140226095754_rename_to_config_group.rb]: we can squash the migrations, but I didn't want to alter your original commit.

File: app/assets/javascripts/class_edit.js [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+function removeConfigGroup(item){
+  var id = $(item).attr('data-group-id');
+  $('#config_group_' + id).removeClass('selected-marker').show();
+  $('#selected_config_group_' + id).children('a').tooltip('hide');
+  $('#selected_config_group_' + id).remove();
+
+  var puppetclasses =  $(item).attr('data-puppetclass_ids')
+  var puppetclass_ids = $.parseJSON(puppetclasses);
+  console.log('trying to remove ' + puppetclass_ids);
```
Comment (ohadlevy) [inline app/assets/javascripts/class_edit.js]: should we keep?

File: app/assets/stylesheets/application.scss [hunk 22]
```diff
@@ -340,3 +345,12 @@ table.table thead .sorting_desc_disabled { background: #f9f9f9 url('sort_desc_di
...
+  color:#aaa;
```
Comment (ohadlevy) [inline app/assets/stylesheets/application.scss]: dont we have a disabled class already ?

File: config/routes.rb (hunk 19 above)
Comment (isratrade) [inline config/routes.rb:151]: I added permissions in the access_permissions.rb.  This is the routes.rb file.

File: app/assets/javascripts/class_edit.js (hunk 21 above)
Comment (isratrade) [inline app/assets/javascripts/class_edit.js]: still more cleaning up to do.

File: app/controllers/hostgroups_controller.rb (hunk 10 above)
Comment (ohadlevy) [inline app/controllers/hostgroups_controller.rb:91]: only one place if you do it in a helper, and this would avoid the need to change the render line everywhere

File: app/controllers/puppetclasses_controller.rb [hunk 23]
```diff
@@ -69,12 +69,12 @@ def get_host_or_hostgroup
...
-        @obj.attributes = params['host'].except!(:puppetclass_ids)
+        @obj.attributes = params['host'].except!(:puppetclass_ids, :config_group_ids)
       elsif params['hostgroup']
         # hostgroup.id is assigned to params['host_id'] by host_edit.js#load_puppet_class_parameters
         @obj = Hostgroup.find(params['host_id'])
         # puppetclass_ids is removed since it causes an insert on hostgroup_classes before form is submitted
```
Comment (ohadlevy) [inline app/controllers/puppetclasses_controller.rb]: ditto, maybe keep the attr_array in a var

File: config/routes.rb (hunk 19 above)
Comment (isratrade) [inline config/routes.rb:151]: Ignore.  I see the test permission failure now.  Strange. I thought it passed locally.

**Comment by ohadlevy** [comment]
looks good, lots of small comments.

I'm missing two things:
1. audits.
2. search on hostgroup/host.

File: app/models/host/managed.rb [hunk 24]
```diff
@@ -297,7 +303,53 @@ def puppetclasses_names
...
+  def puppetclass_ids
+    classes.reorder('').pluck('puppetclasses.id')
+  end
+
+  def individual_puppetclasses
+    puppetclasses - classes_in_groups
+  end
+
+  def classes_in_groups
+    # TODO - REFACTOR - BAD
```
Comment (ohadlevy) [inline app/models/host/managed.rb]: # TODO - REFACTOR - BAD
maybe even

``` ruby
ids = ConfigGroupClass.where(:config_group_id => config_groups).pluck(:puppetclass_id)
```

File: app/models/host/managed.rb [hunk 25]
```diff
@@ -297,7 +303,53 @@ def puppetclasses_names
...
+    ids = []
+    self.config_groups.each do |config_group|
+      ids += config_group.config_group_classes.pluck(:puppetclass_id)
+    end
+    Puppetclass.where(:id => ids.uniq) - parent_classes
```
Comment (ohadlevy) [inline app/models/host/managed.rb]: and in oneliner

``` ruby
 Puppetclass.where(:id => ConfigGroupClass.where(:config_group_id => config_group_ids).select(:puppetclass_id))
```

File: app/models/hostgroup.rb [hunk 26]
```diff
@@ -78,14 +84,61 @@ def diskLayout
...
   end
 
-  def classes
-    Puppetclass.joins(:hostgroups).where(:hostgroups => {:id => path_ids})
+  def individual_puppetclasses
+    puppetclasses - classes_in_groups
+  end
+
+  def classes(env = nil)
+    # TODO - REFACTOR - BAD
```
Comment (ohadlevy) [inline app/models/hostgroup.rb]: ditto

File: app/models/hostgroup.rb [hunk 27]
```diff
@@ -78,14 +84,61 @@ def diskLayout
...
+      ids += Puppetclass.joins(:hostgroups).where(:hostgroups => {:id => hostgroup.path_ids}).pluck(:id)
+      ids += hostgroup.config_group_classes.pluck(:puppetclass_id)
+    end
+    return (Puppetclass.where(:id => ids.uniq)) unless env || environment
+    return env.puppetclasses.where('puppetclasses.id' => ids.uniq) if env
+    return environment.puppetclasses.where('puppetclasses.id' => ids.uniq) if environment
+  end
+
+  def classes_in_groups
+    # TODO - REFACTOR - BAD
```
Comment (ohadlevy) [inline app/models/hostgroup.rb]: ditto

Comment (ohadlevy) [inline app/models/hostgroup.rb]: DRY, maybe use a scope if it make sense?

File: app/models/hostgroup.rb [hunk 28]
```diff
@@ -78,14 +84,61 @@ def diskLayout
...
   end
 
+  def parent_classes
+    return [] if is_root?
+    parent.classes(self.environment)
+  end
+
+  def parent_class_ids
+    return [] if is_root?
+    parent.classes.map(&:id)
```
Comment (ohadlevy) [inline app/models/hostgroup.rb]: pluck?

**Comment by ohadlevy** [comment]
could you please add a few more tests to all of the class retrievals ?

File: app/views/config_groups/edit.html.erb (hunk 1 above)
Comment (isratrade) [inline app/views/config_groups/edit.html.erb:3]: Yes, partial exists.
https://github.com/isratrade/foreman/blob/classes/app/views/config_groups/_form.html.erb

File: app/models/hostgroup.rb (hunk 28 above)
Comment (isratrade) [inline app/models/hostgroup.rb]: `pluck` causes an error

```
ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list
LINE 1: ... 236, 70, 15, 72, 6, 2, 21, 407, 401, 5) ORDER BY puppetclas...
                                                             ^
: SELECT DISTINCT puppetclasses.id FROM "puppetclasses" INNER JOIN "environment_classes" ON "puppetclasses"."id" = "environment_classes"."puppetclass_id" WHERE "environment_classes"."environment_id" = 1 AND "puppetclasses"."id" IN (54, 229, 236, 70, 15, 72, 6, 2, 21, 407, 401, 5) ORDER BY puppetclasses.name
```

File: app/views/config_groups/index.html.erb (hunk 15 above)
Comment (isratrade) [inline app/views/config_groups/index.html.erb]: I changed the list of puppet classes to count, so there is no issue with max-width.

File: app/assets/stylesheets/application.scss (hunk 22 above)
Comment (isratrade) [inline app/assets/stylesheets/application.scss]: I changed the class to 'gray' which is already defined rather than creating a new class

File: app/controllers/puppetclasses_controller.rb (hunk 23 above)
Comment (isratrade) [inline app/controllers/puppetclasses_controller.rb]: @ohadlevy, why create a new var for `.attributes` when `obj` is returned by the method?

Comment (ohadlevy) [inline app/controllers/puppetclasses_controller.rb]: @isratrade simply because you define the attributes (:puppetclass_ids, :config_group_ids) in line 72 and line 77

Comment (isratrade) [inline app/controllers/puppetclasses_controller.rb]: @ohadlevy, I'm not defining attributes (:puppetclass_ids, :config_group_ids).  I'm removing these keys from the params[:hostgroup] hash using `except!`

File: app/controllers/puppetclasses_controller.rb [hunk 29]
```diff
@@ -68,13 +68,13 @@ def get_host_or_hostgroup
           @obj      = @obj.becomes(Host::Managed)
           @obj.type = "Host::Managed"
         end
-        # puppetclass_ids is removed since it causes an insert on host_classes before form is submitted
-        @obj.attributes = params['host'].except!(:puppetclass_ids)
+        # puppetclass_ids and config_group_ids are removed since they causes an insert on host_classes and config_group_classes before form is submitted
+        @obj.attributes = params['host'].except!(:puppetclass_ids, :config_group_ids)
```
Comment (ohadlevy) [inline app/controllers/puppetclasses_controller.rb:72]: sometimes an example is easier:

``` ruby
ignored_parameters = [:puppetclass_ids, :config_group_ids]
@obj.attributes = params['host'].except!(ignored_parameters)
...
@obj.attributes = params['hostgroup'].except!(ignored_parameters)
```

**Comment by ares** [comment]
I created three nested hostgroups **a**, **a/b**, **a/b/c**. When I assign config role to **a** I can see it in **a/b** (without ability to remove assignment) but I don't see it in **a/b/c**. I think the inheritance either shouldn't work at all or for all descendants. Also if I first assign config role to **a/b** then the same to **a** I see it twice in **a/b** (which is ok, would be nice I could see from where it comes) but I can't remove any of those. I should be able to remove the one I added specifically for **a/b**.

EDIT: actually I'm not sure whether inheritance should be applied here. It makes things unclear. We are adding this because we need to address n:m assignment. If we want to assign some class to the hostgroup and subhostgroups we can assign class to parent hostgroup directly.

File: app/controllers/puppetclasses_controller.rb (hunk 29 above)
Comment (isratrade) [inline app/controllers/puppetclasses_controller.rb:72]: @ohadlevy, that makes sense :)

**Comment by stbenjam** [comment]
This is really cool.

Just a small nitpick, can you make the class counter sortable so it's consistent with the UI?  And add a hosts counter that links to a search on hosts using the config group? 

I can come back and add it later if there's not time for 1.5, it'd just be nice since all numerical columns are sortable now.

**Comment by GregSutcliffe** [comment]
@isratrade is there no concept of environments within a config group? If I have:

```
environments/
  prod/
    modules/
      ntp/
  dev/
    modules/
      apache/
```

These are two entirely independant and separate things, yet I can create a config group which contains both. More confusingly, I can add this config group to a host in prod and it will _actually_ add apache to the host, when in fact that's not possible, and will cause a puppet error.

I think the groups should still be scoped by environment, or perhaps we need to think about this some more?

**Comment by ohadlevy** [comment]
On Apr 7, 2014 6:19 PM, "Greg Sutcliffe" notifications@github.com wrote:

> @isratrade is there no concept of environments within a config group? If
> I have:
> 
> environments/
>   prod/
>     modules/
>       ntp/
>   dev/
>     modules/
>       apache/
> 
> These are two entirely independant and separate things, yet I can create
> a config group which contains both. More confusingly, I can add this config
> group to a host in prod and it will actually add apache to the host, when
> in fact that's not possible, and will cause a puppet error.

It was by design to not scope it to an environment, however it should first
indicate/include only classes that belong to the current env and secondly
filter them in the future (eg if you added a class to a group later on)
based on an existing setting that wet have to do that... Maybe we should
force it to true if you are using this feature

**Comment by isratrade** [comment]
@stbenjam, thanks for the feedback.  I can add this.

**Comment by isratrade** [comment]
@GregSutcliffe, if a puppetclass is not in the environment of a host/hostgroup, note that it is grayed out and has a tool tip "not in environment".  This means that a config group can have 4 puppetclasses but when it's added to a host/hostgroup, only 3 puppetclasses are actually included.

**Comment by isratrade** [comment]
added search by config_group for hosts and hostgroups

**Comment by isratrade** [comment]
@GregSutcliffe, what is the value of your `Setting['remove_classes_not_in_environment']`
If you change it to `true` do you get back you expect in the ENC output

**Comment by GregSutcliffe** [comment]
@isratrade it's 'false' which I assume is the default since I didn't touch it. Changing it to true does indeed fix the ENC yaml, so I suggest this default to 'true'.

I don't see any mention of "not in environment" anywhere on the Edit page, for either value of that Setting. Nor are the classes in the other environment greyed out. This is on Firefox 27.0.1

**Comment by GregSutcliffe** [comment]
@ohadlevy if it was delibrate, fair enough - in that case, I'm just flagging that right now, I applied the patch and tried the feature and it broke things. Users won't like that :)

**Comment by isratrade** [comment]
@GregSutcliffe, what broke? This PR shouldn't break anything if merged as config groups need to be manually defined.  If none are, then this PR doesn't change anything.

**Comment by isratrade** [comment]
@GregSutcliffe, please review PR #1360. Thanks!

**Comment by GregSutcliffe** [comment]
@isratrade by "broke" I mean my install was passing a list of classes to Puppet that is not possible to apply (because they're in two different environments) with no warning to the user. Neither the config_group edit page nor the host edit page showed me the "not in environment" grey-out that you mentioned above, so I was not warned that I'd created an incompatible set of classes. Foreman itself is fine with this, but the next puppet run will produce errors, because of the missing classes.

I've had a look at 1360, looks OK. I'll test it shortly and comment. We should also add a migration to this PR to set that Setting to true, as most people aren't going to want to be allowed to create bad config groups.

File: app/controllers/puppetclasses_controller.rb [hunk 30]
```diff
@@ -62,19 +62,19 @@ def get_host_or_hostgroup
...
+      ignored_parameters = [:puppetclass_ids, :config_group_ids]
       if params['host']
         @obj = Host::Base.find(params['host_id'])
         unless @obj.kind_of?(Host::Managed)
           @obj      = @obj.becomes(Host::Managed)
           @obj.type = "Host::Managed"
         end
-        # puppetclass_ids is removed since it causes an insert on host_classes before form is submitted
-        @obj.attributes = params['host'].except!(:puppetclass_ids)
+        @obj.attributes = params['host'].except!(*ignored_parameters)
```
Comment (isratrade) [inline app/controllers/puppetclasses_controller.rb]: I assume the Ruby 1.8.7 fails were due to this splat operator. I'll change it back.

**Comment by isratrade** [comment]
@GregSutcliffe, you need to click on the config group and the upperclassmen will open up underneath it.  If a class is not in the current environment, then it should be grayed-out with a tooltip that says "not in environment"

**Comment by ares** [comment]
it seems that in last version I can't diassociate config roles from hostgroup, I click remove, they appear on right side, then I hit submit and I see success notification but when I get back it's still assigned

**Comment by GregSutcliffe** [comment]
@isratrade i do see it now - I had expanded the Config Group, but the grey of 'not-in-environment' and the black of 'ok' are so similar that I totally missed it. Can we make this a much lighter grey? Also an icon would be good, to indicate a problem to the user.

It certainly makes more sense now - looking good apart from needing the change to 'remove_classes_not_in_environment' and these cosmetic things :)

**Comment by GregSutcliffe** [comment]
@ares I can't replicate that, works fine here. maybe @isratrade already fixed it.

@isratrade found a small issue - try this sequence:

1) edit a host or hostgroup with no assigned config groups
2) add a class directly (eg ntp) - it appears in the included classes list
3) add a config group which includes that class (eg a group containing ntp)
4) click the remove button next to the class
-> class is removed from the list even though it's in the config group

What i'd expect is that the remove button next to the class is disabled when you add a config group with the class in it

EDIT: forgot to say, this is entirely cosmetic - if you save it like that (with the class missing from the list) and re-edit, it will be correctly on the list. Just a JS issue :)

**Comment by isratrade** [comment]
thanks @ares, I fixed it.  It was introduced after I re-factored so that selected, inherited, or unselected config groups use the same partial `_config_group.html`

**Comment by isratrade** [comment]
@GregSutcliffe, I thought about this use case you brought up. After a hostgroup is save for an included class and a config group that contains the same class, there is no "Minus" on the class as the config group has priority.  You need to remove the config group in order to remove the class.  I assume that the JS when adding a config group should remove the "minus" on the class, right?

**Comment by isratrade** [comment]
updated with hostgroup unit tests. still need to do host unit tests on new methods introduced

**Comment by isratrade** [comment]
not sure about the sqlite test failure on sort, any ideas?

File: app/assets/javascripts/class_edit.js [hunk 31]
```diff
@@ -0,0 +1,130 @@
+function filter_puppet_classes(item){
```
Comment (domcleal) [inline app/assets/javascripts/class_edit.js:1]: Add this to config/environments/production.rb, the list of JS files.

**Comment by isratrade** [comment]
all tests pass!  which nitpicks / DRY re-factoring are crucial before merge?

**Comment by ares** [comment]
@isratrade have you looked at the inheritance issue I described here? https://github.com/theforeman/foreman/pull/1217#issuecomment-39720026
I didn't see any reply to that

**Comment by isratrade** [comment]
@ares, I'm not sure about inheritance. Config groups are not inherited. There is only the element of defining a config group on a hostgroup (which has inheritance).  But this shouldn't be any different than defining a domain on a hostgroup.

**Comment by ares** [comment]
@isratrade this may be some issue on my setup since noone else seems to hit it but... when I have three nested hostgroups a -> a/b -> a/b/c and I assign config group to a, it's automatically assigned to a/b and not assigned to a/b/c. Maybe the automatic assignment to a/b was not intended at all.

**Comment by isratrade** [comment]
@GregSutcliffe, I updated the color from gray to light gray (#999) used by bootstrap

**Comment by isratrade** [comment]
@ares, it should be assigned to a/b/c.  I will look into it.

**Comment by isratrade** [comment]
[test] rundeck failure again

**Comment by isratrade** [comment]
@ares, I fixed the a/b/c issue in a separate commit to be squashed after review

**Comment by isratrade** [comment]
updated with sortable counter columns on config_groups#index with links to hosts, hostgroups, and puppetclasses per @stbenjam request

**Comment by ohadlevy** [comment]
merged! thanks @isratrade
